The senior Instacart officers have given you a new data set of customer information to go along with your product and order data. In part 1 of the task, you’ll need to incorporate this new data set into your project. In part 2, you’ll create some visualizations, conduct some exploratory analysis, and begin wrapping up everything you’ve done in this Achievement 

In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
#Read in dataset

df=pd.read_csv(r"C:\Users\senoe\OneDrive\SEN Docs\Data Sci Job Applications\CareerFoundry\Datasets\original data\customers.csv", index_col=False)

In [3]:
#Check df for basic information

df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [4]:
#Rename columns for better clarity

df.rename(columns = {'First Name':'first_name', 'Surnam':'surname',
                              'Gender':'gender', 'STATE':'state', 'Age':'age', 'n_dependants':'num_dependants'}, inplace = True)

In [5]:
#Check the renamed columns

df.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_dependants fam_status  income  
0               3    married  165665  
1               0     single   59285  
2               2    married   99568  
3               0     single   42049  
4               1    married   40374

Checking the date_joined column to see if all the dates are the same.  If yes, then I will drop the column

In [6]:
#checking column to see if all the dates are the same

df['date_joined'].head(100)

0     1/1/2017
1     1/1/2017
2     1/1/2017
3     1/1/2017
4     1/1/2017
        ...   
95    1/1/2017
96    1/1/2017
97    1/1/2017
98    1/1/2017
99    1/1/2017
Name: date_joined, Length: 100, dtype: object

In [7]:
df['date_joined'].tail(100)

206109    4/1/2020
206110    4/1/2020
206111    4/1/2020
206112    4/1/2020
206113    4/1/2020
            ...   
206204    4/1/2020
206205    4/1/2020
206206    4/1/2020
206207    4/1/2020
206208    4/1/2020
Name: date_joined, Length: 100, dtype: object

At this point I will not drop any column as it seems like all the data has some relevancy.

In [8]:
#Checking the overall info for the df

df.describe()

user_id            age  num_dependants         income
count  206209.000000  206209.000000   206209.000000  206209.000000
mean   103105.000000      49.501646        1.499823   94632.852548
std     59527.555167      18.480962        1.118433   42473.786988
min         1.000000      18.000000        0.000000   25903.000000
25%     51553.000000      33.000000        0.000000   59874.000000
50%    103105.000000      49.000000        1.000000   93547.000000
75%    154657.000000      66.000000        3.000000  124244.000000
max    206209.000000      81.000000        3.000000  593901.000000

In [9]:
#checking df for missing values

df.isnull().sum()

user_id               0
first_name        11259
surname               0
gender                0
state                 0
age                   0
date_joined           0
num_dependants        0
fam_status            0
income                0
dtype: int64

There are 11259 missing values in the first name column.  

In [10]:
#check the dtypes for df

df.dtypes

user_id            int64
first_name        object
surname           object
gender            object
state             object
age                int64
date_joined       object
num_dependants     int64
fam_status        object
income             int64
dtype: object

The df is comprised of two types of values:

1. object(str or mixed) = Text or mixed numeric and non-numeric values)  
2. int64 (int) = Integer numbers

In [11]:
df['first_name'].fillna("None", inplace = True)

Will impute rows with missing values from df in 'first_name' column.

In [12]:
#checking df for missing values

df.isnull().sum()

user_id           0
first_name        0
surname           0
gender            0
state             0
age               0
date_joined       0
num_dependants    0
fam_status        0
income            0
dtype: int64

In [13]:
df.head(100)

user_id  first_name   surname  gender       state  age date_joined  \
0     26711     Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1     33890    Patricia      Hart  Female  New Mexico   36    1/1/2017   
2     65803     Kenneth    Farley    Male       Idaho   35    1/1/2017   
3    125935    Michelle     Hicks  Female        Iowa   40    1/1/2017   
4    130797         Ann   Gilmore  Female    Maryland   26    1/1/2017   
..      ...         ...       ...     ...         ...  ...         ...   
95    86389     Stephen   Murillo    Male     Vermont   69    1/1/2017   
96   102956       Kathy  Mitchell  Female    Missouri   78    1/1/2017   
97   180919  Jacqueline     Roach  Female    Nebraska   21    1/1/2017   
98    74155       Karen      Berg  Female     Alabama   33    1/1/2017   
99    96166        None   Oconnor    Male    Oklahoma   51    1/1/2017   

    num_dependants                        fam_status  income  
0                3                           married  165665  
1                0                            single   59285  
2                2                           married   99568  
3                0                            single   42049  
4                1                           married   40374  
..             ...                               ...     ...  
95               3                           married   66800  
96               1                           married   96314  
97               3  living with parents and siblings   56831  
98               3                           married   95285  
99               1                           married  155673  

[100 rows x 10 columns]

In [14]:
#Checking the overall info for the df

df.describe()

user_id            age  num_dependants         income
count  206209.000000  206209.000000   206209.000000  206209.000000
mean   103105.000000      49.501646        1.499823   94632.852548
std     59527.555167      18.480962        1.118433   42473.786988
min         1.000000      18.000000        0.000000   25903.000000
25%     51553.000000      33.000000        0.000000   59874.000000
50%    103105.000000      49.000000        1.000000   93547.000000
75%    154657.000000      66.000000        3.000000  124244.000000
max    206209.000000      81.000000        3.000000  593901.000000

In [15]:
# Selecting duplicate rows except first
# occurrence based on all columns
df_dup = df[df.duplicated(keep=False)]


In [16]:
# Print the resultant Dataframe
df_dup

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, num_dependants, fam_status, income]
Index: []

Used the duplicated() function to find duplicate values in the DataFrame. Based on Output there aren't any duplicated in the df.

In [17]:
#list columns with mixed types in df

df.loc[:, df.applymap(type).nunique().gt(1)]

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]

[206209 rows x 0 columns]

In [18]:
df.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_dependants fam_status  income  
0               3    married  165665  
1               0     single   59285  
2               2    married   99568  
3               0     single   42049  
4               1    married   40374

In [19]:
df.shape

(206209, 10)

In [20]:
#read in pickle file

ords_prods_merge = pd.read_pickle(r'C:\Users\senoe\OneDrive\SEN Docs\Data Sci Job Applications\CareerFoundry\Datasets\Prepared data\orders_products_merge.pkl')

In [21]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2398795        1    prior             2                       3   
2    473747        1    prior             3                       3   
3   2254736        1    prior             4                       4   
4    431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                       0         196                  1   
1                  7                      15         196                  1   
2                 12                      21         196                  1   
3                  7                      29         196                  1   
4                 15                      28         196                  1   

   reordered  ... aisle_id department_id  prices    price_range_loc  \
0          0  ...       77             7     9.0  Mid_range product   
1          1  ...       77             7     9.0  Mid_range product   
2          1  ...       77             7     9.0  Mid_range product   
3          1  ...       77             7     9.0  Mid_range product   
4          1  ...       77             7     9.0  Mid_range product   

      busiest_day busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy         Fewest orders        10  New customer   
1  Regularly busy         Fewest orders        10  New customer   
2  Regularly busy        Highest orders        10  New customer   
3      Least busy         Fewest orders        10  New customer   
4      Least busy        Highest orders        10  New customer   

   loyalty_flag_price             order_flag  
0         Low spender      Frequent customer  
1         Low spender       Regular customer  
2         Low spender  Non-frequent customer  
3         Low spender  Non-frequent customer  
4         Low spender  Non-frequent customer  

[5 rows x 22 columns]

In [22]:
#Joining DataFrames 

df_combined = ords_prods_merge.merge(df, on = 'user_id')

In [23]:
#checking thecombined df

df_combined.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2398795        1    prior             2                       3   
2    473747        1    prior             3                       3   
3   2254736        1    prior             4                       4   
4    431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                       0         196                  1   
1                  7                      15         196                  1   
2                 12                      21         196                  1   
3                  7                      29         196                  1   
4                 15                      28         196                  1   

   reordered  ...             order_flag first_name  surname  gender    state  \
0          0  ...      Frequent customer      Linda   Nguyen  Female  Alabama   
1          1  ...       Regular customer      Linda   Nguyen  Female  Alabama   
2          1  ...  Non-frequent customer      Linda   Nguyen  Female  Alabama   
3          1  ...  Non-frequent customer      Linda   Nguyen  Female  Alabama   
4          1  ...  Non-frequent customer      Linda   Nguyen  Female  Alabama   

  age date_joined num_dependants  fam_status income  
0  31   2/17/2019              3     married  40423  
1  31   2/17/2019              3     married  40423  
2  31   2/17/2019              3     married  40423  
3  31   2/17/2019              3     married  40423  
4  31   2/17/2019              3     married  40423  

[5 rows x 31 columns]

In [24]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_the_week  int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  Int64  
 7   product_id              int64  
 8   add_to_cart_order       int64  
 9   reordered               int64  
 10  _merge                  object 
 11  product_name            object 
 12  aisle_id                int64  
 13  department_id           int64  
 14  prices                  float64
 15  price_range_loc         object 
 16  busiest_day             object 
 17  busiest_period_of_day   object 
 18  max_order               int64  
 19  loyalty_flag            object 
 20  loyalty_flag_price      object 
 21  order_flag              objec

In [25]:
df_combined.shape

(32404859, 31)

In [26]:
#Create a path
path_new = r'C:\Users\senoe\OneDrive\SEN Docs\Data Sci Job Applications\CareerFoundry\Datasets'

In [27]:
# Export data to pkl

df_combined.to_pickle(os.path.join(path_new, 'Prepared data', 'orders_products_merge_comined_2.pkl'))